File: C++/NextPermutation.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
    void nextPermutation(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int len = num.size();
		if(len <= 1)
			return;
		int i = len-2;
		while(i>=0 && num[i] >= num[i+1])
			i--;
		if(i == -1)
		{
			reverse(num.begin(), num.end());
			return;
		}
		int j = len-1;
		while(num[j] <= num[i])
			j--;
		swap(num[i], num[j]);
		reverse(num.begin()+i+1,num.end());
    }
};
int main()
{
	vector<int> num;
	num.push_back(1);
	num.push_back(1);
	Solution sol;
	sol.nextPermutation(num);
	getchar();
}

File: JAVA/NextPermutation.java

class Solution {
    public void nextPermutation(int[] num) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(num == null || num.length == 0)
        	return;
        int pos1 = num.length-1;
        while(pos1 >= 1 && num[pos1] <= num[pos1-1])
        {
        	pos1--;
        }
        if(pos1 == 0)
        {
        	for(int i = 0; i < num.length/2; i++)
        	{
        		int t = num[i];
        		num[i] = num[num.length-i-1];
        		num[num.length-1-i] = t;
        	}
        	return;
       	}
        pos1--;
        int pos2 = num.length-1;
        while(pos2 > pos1 && num[pos2] <= num[pos1])
        	pos2--;
        int t = num[pos1];
        num[pos1] = num[pos2];
        num[pos2] = t;
        for(int i = pos1+1; i < (num.length + pos1 + 1)/2; i++)
        {
    		t = num[i];
    		num[i] = num[num.length-i + pos1];
    		num[num.length-i+pos1] = t;
        }
    }
}
public class NextPermutation {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
